akka-projection-grpc/src/main/scala/akka/projection/grpc/producer/scaladsl/EventProducer.scala

```diff
@@ -61,25 +67,44 @@ object EventProducer {
    */
   @ApiMayChange
   final class Transformation private (
-      val mappers: Map[Class[_], Any => Future[Option[Any]]],
-      val orElse: Any => Future[Option[Any]]) {
+      private[akka] val mappers: Map[Class[_], Transformation.Mapper[Any, Any]],
+      private[akka] val orElse: Transformation.Mapper[Any, Any]) {
+    import Transformation._
+
+    /**
+     * @param f A function that is fed each event, and the possible additional metadata
+     */
+    def registerLowLevelMapper[A: ClassTag, B](m: Mapper[A, B]): Transformation = {
```

Comment:
what do you think about having it for the `EventEnvelope` instead?

Reply:
That was my original idea, maybe that is better because of the existing type and no need for a SAM to make the two params clear like with this solution.

Reply:
and we give access to the additional envelope properties

Reply:
Ah, yes, now that I did it, I realise the pid needed in Kalix is only available through the envelope, not the metadata. Changed in [a0f6880](https://github.com/akka/akka-projection/pull/764/commits/a0f68806d469e7c20bc3e67e41d14059f9cc7065)